text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,391 @@
+We want to find a generic mechanism for running these tests, without having to cherry-pick use-cases or versions. 
+
+The main design idea is to introduce new types of integration tests into our pre-release suite. Those tests will run various permutations of `(CLI X Framework)`, and validate the expected result in each permutation.
+
+To make it generic, the tests will run the entire current integration tests suite, and simply inject different versions each time.
+
+An important note is that when testing compatibility to older versions (be it CLI or Framework), we cannot run the *latest* integration test suite, as it may contain tests for new features and bug fixes, which were not implemented in older versions.
+
+Therefore, the entire design relies on us having the ability to:
+
+- *Run **old** integration tests using a **new** CLI version and an **old** Framework version.*
+- *Run **old** integration tests using an **old** CLI version and a **new** Framework version.*
+- *Mark a Framework version as incompatible with the latest published CLI version.*
+
+    - The tests should support an ability to specify versions who are not compatible, since the expected output would be different.
+
+# Detailed Design
+
+As mentioned, everything is based on our ability to run integration tests with various version permutations. We introduce a new function, `run_integration_tests_perm` which magically does what we need, as well as some useful constants.
+
+The design is best explained by simply looking at the proposed code changes. Most of what you need to know is in the inline comments of the code.
+
+**[common.bash](https://github.com/aws/aws-cdk/blob/master/packages/aws-cdk/test/integ/cli/common.bash)**
+
+```bash
+function bump_minor() {
+
+    version=$1
+
+    # Accept a version and bump the minor version.
+}
+
+# This is the latest published version.
+LATEST_PUBLISHED_VERSION=$(node -p 'require("package.json").version')
+
+# This is the soon to be published version.
+NEXT_PUBLISHED_VERSION="$(bump_minor ${LATEST_PUBLISHED_VERSION})"
+
+function run_integration_tests_perm() {
+
+    integration_tests_version=1
+    framework_version=$2
+    cli_version=$3
+    expected_error=$4
+
+    # Note that this function supports installing from source by using the "SOURCE" string 
+    # as the version identifier.
+
+    # Checkout the relevant integration tests
+    # Install framework version
+    # Install CLI version
+    # Inject versions to integration tests
+    # Run tests (should either pass or print the expected error)
+
+}
+
+```
+
+We now create a new file, which is responsible for validating the compatibility of new framework versions against old cli versions. This test will have the following assertions:
+
+- Compatible versions must pass all integration tests.
+- Incompatible versions must print the expected error message.
```

Comment:
New CLI must always be compatible, so I am not exactly clear what are "incompatible" versions.

Reply:
The context of these statements are regarding framework compatibility.

```text
We now create a new file, which is responsible for validating the compatibility of new framework versions against old cli versions. This test will have the following assertions:

- Compatible versions must pass all integration tests.
- Incompatible versions must print the expected error message.


```